File: go/internal/coordinator/probe.go
package coordinator

import (
	"context"
	"net/http"
	"sync"
	"time"
)

// nodeProbe tracks health of each configured storage node. A background
// goroutine pings /healthz periodically; pickNodes / streamBlock consult
// the live status so a downed node is skipped fast instead of timing out.
type nodeProbe struct {
	interval time.Duration
	timeout  time.Duration
	client   *http.Client
	nodes    []string

	mu      sync.RWMutex
	up      map[string]bool
	lastErr map[string]string

	cancel context.CancelFunc
	done   chan struct{}

	onChange func(node string, up bool)
}

func newNodeProbe(nodes []string, interval, timeout time.Duration, onChange func(string, bool)) *nodeProbe {
	p := &nodeProbe{
		interval: interval,
		timeout:  timeout,
		client:   &http.Client{Timeout: timeout},
		nodes:    nodes,
		up:       map[string]bool{},
		lastErr:  map[string]string{},
		onChange: onChange,
	}
	// optimistic start: assume up so we don't reject all writes for the first tick
	for _, n := range nodes {
		p.up[n] = true
	}
	return p
}

func (p *nodeProbe) start() {
	if p.interval <= 0 {
		return
	}
	ctx, cancel := context.WithCancel(context.Background())
	p.cancel = cancel
	p.done = make(chan struct{})
	go p.loop(ctx)
}

func (p *nodeProbe) stop() {
	if p.cancel != nil {
		p.cancel()
		<-p.done
	}
}

func (p *nodeProbe) isUp(node string) bool {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.up[node]
}

func (p *nodeProbe) loop(ctx context.Context) {
	defer close(p.done)
	// first round: immediate
	p.probeAll(ctx)
	t := time.NewTicker(p.interval)
	defer t.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-t.C:
			p.probeAll(ctx)
		}
	}
}

func (p *nodeProbe) probeAll(ctx context.Context) {
	var wg sync.WaitGroup
	for _, n := range p.nodes {
		wg.Add(1)
		go func(node string) {
			defer wg.Done()
			p.probeOne(ctx, node)
		}(n)
	}
	wg.Wait()
}

func (p *nodeProbe) probeOne(ctx context.Context, node string) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, node+"/healthz", nil)
	if err != nil {
		p.set(node, false, err.Error())
		return
	}
	resp, err := p.client.Do(req)
	if err != nil {
		p.set(node, false, err.Error())
		return
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		p.set(node, false, resp.Status)
		return
	}
	p.set(node, true, "")
}

func (p *nodeProbe) set(node string, up bool, errStr string) {
	p.mu.Lock()
	prev, known := p.up[node]
	p.up[node] = up
	p.lastErr[node] = errStr
	p.mu.Unlock()
	if p.onChange != nil && (!known || prev != up) {
		p.onChange(node, up)
	}
}

File: ts/src/coordinator/probe.ts
// Background health probe for storage nodes. Mirrors go/internal/coordinator/probe.

export interface ProbeOptions {
  nodes: string[];
  intervalMs: number;
  timeoutMs: number;
  onChange?: (node: string, up: boolean) => void;
}

export class NodeProbe {
  private up = new Map<string, boolean>();
  private timer: NodeJS.Timeout | null = null;

  constructor(private opts: ProbeOptions) {
    // optimistic start so we don't reject all writes during the first tick
    for (const n of opts.nodes) this.up.set(n, true);
  }

  start(): void {
    if (this.opts.intervalMs <= 0) return;
    void this.probeAll();
    this.timer = setInterval(() => void this.probeAll(), this.opts.intervalMs);
    // unref so this doesn't keep the event loop alive in tests
    this.timer.unref?.();
  }

  stop(): void {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
  }

  isUp(node: string): boolean {
    return this.up.get(node) ?? false;
  }

  private async probeAll(): Promise<void> {
    await Promise.all(this.opts.nodes.map((n) => this.probeOne(n)));
  }

  private async probeOne(node: string): Promise<void> {
    const ctrl = new AbortController();
    const t = setTimeout(() => ctrl.abort(), this.opts.timeoutMs);
    let up = false;
    try {
      const r = await fetch(node + "/healthz", { signal: ctrl.signal });
      up = r.status === 200;
      // drain
      await r.arrayBuffer().catch(() => undefined);
    } catch {
      up = false;
    } finally {
      clearTimeout(t);
    }
    const prev = this.up.get(node);
    this.up.set(node, up);
    if (prev !== up) this.opts.onChange?.(node, up);
  }
}
